Sources/ActorSingletonPlugin/ActorSingletonPlugin.swift

```diff
@@ -22,48 +23,70 @@ import DistributedActorsConcurrencyHelpers
-/// `_ActorRef` is in reality a proxy which handle situations where the singleton is shifted to different nodes.
+/// its allocation. On candidate nodes where the singleton might run, use `ClusterSystem.singleton.proxy(type:name:system:props:factory)`
+/// to define the distributed actor. Otherwise, call `ClusterSystem.singleton.proxy(type:name:)` to obtain a distributed actor. The returned
+/// `DistributedActor` is in reality a proxy that handles situations where the singleton might get relocated to different nodes.
 ///
 /// - Warning: Refer to the configured `AllocationStrategy` for trade-offs between safety and recovery latency for
 ///            the singleton allocation.
 /// - SeeAlso: The `ActorSingleton` mechanism is conceptually similar to Erlang/OTP's <a href="http://erlang.org/doc/design_principles/distributed_applications.html">`DistributedApplication`</a>,
 ///            and <a href="https://doc.akka.io/docs/akka/current/cluster-singleton.html">`ClusterSingleton` in Akka</a>.
-public final class ActorSingletonPlugin {
-    private var singletons: [String: BoxedActorSingleton] = [:]
-    private let singletonsLock = Lock()
-
-    public init() {}
-
-    // FIXME: document that may crash, it may right?
-    func ref<Message: Codable>(of type: Message.Type, settings: ActorSingletonSettings, system: ClusterSystem, props: _Props? = nil, _ behavior: _Behavior<Message>? = nil) throws -> _ActorRef<Message> {
-        try self.singletonsLock.withLock {
-            if let existing = self.singletons[settings.name] {
-                guard let proxy = existing.unsafeUnwrapAs(Message.self).proxy else {
-                    fatalError("Singleton [\(settings.name)] not yet initialized")
-                }
-                return proxy
-            }
-
-            let singleton = ActorSingleton<Message>(settings: settings, props: props, behavior)
-            try singleton.startAll(system)
-            self.singletons[settings.name] = BoxedActorSingleton(singleton)
+public actor ActorSingletonPlugin {
+    private var singletons: [String: (proxied: ActorID, proxy: any AnyActorSingletonProxy)] = [:]
+
+    public func proxy<Act>(
+        of type: Act.Type,
+        settings: ActorSingletonSettings,
+        system: ClusterSystem,
+        props: _Props = _Props(),
+        _ factory: ((ClusterSystem) async throws -> Act)? = nil
+    ) async throws -> Act
+        where Act: DistributedActor,
+        Act.ActorSystem == ClusterSystem
+    {
+        if let existingID = self.singletons[settings.name]?.proxied {
+            return try Act.resolve(id: existingID, using: system)
+        }
 
-            guard let proxy = singleton.proxy else {
-                fatalError("Singleton[\(settings.name)] not yet initialized")
-            }
+        // Spawn the proxy for the singleton (one per singleton per node)
+        let proxy = try await _Props.$forSpawn.withValue(_Props._wellKnownActor(name: "singletonProxy-\(settings.name)")) {
```

Comment:
TODO: rework this entire spawning / props pattern -- I'll work on this as soon as In can this week